test_runner/src/main/kotlin/ftl/reports/outcome/TestOutcomeContext.kt

```diff
@@ -0,0 +1,39 @@
+package ftl.reports.outcome
+
+import com.google.api.services.testing.model.TestMatrix
+import com.google.api.services.testing.model.ToolResultsExecution
+import com.google.api.services.toolresults.model.Environment
+import com.google.api.services.toolresults.model.Step
+import ftl.gc.GcToolResults
+import ftl.util.timeoutToSeconds
+
+data class TestOutcomeContext(
+    val matrixId: String,
+    val projectId: String,
+    val environments: List<Environment>,
+    val steps: List<Step>,
+    val testTimeout: Long
+)
+
+fun TestMatrix.fetchTestOutcomeContext() = getToolResultsIds().let { ids ->
+    TestOutcomeContext(
+        projectId = projectId,
+        matrixId = testMatrixId,
+        environments = GcToolResults.listAllEnvironments(ids),
+        steps = GcToolResults.listAllSteps(ids),
+        testTimeout = testTimeout()
+    )
+}
+
+private fun TestMatrix.getToolResultsIds(): ToolResultsExecution = ToolResultsExecution()
+    .setProjectId(projectId)
+    .setHistoryId(resultStorage?.toolResultsExecution?.historyId ?: throw BadMatrixError())
```

Comment:
I think `BadMatrixErrors` are not handled and will print ugly stacktrace to the user 😢

Reply:
Do you have an idea how this [error](https://github.com/Flank/gcloud_cli/blob/137d864acd5928baf25434cf59b0225c4d1f9319/google-cloud-sdk/lib/googlecloudsdk/api_lib/firebase/test/exceptions.py#L35) is mapped to exit code on gcloud? 😄

Reply:
TBH -- no, I've look at `gcloud` source code but found nothing (maybe it's handled at all ? )
Based on https://firebase.google.com/docs/test-lab/android/command-line#script_exit_codes
I would guess it's `15`, what do you think?

Reply:
I have found nothing in the source code too. `15` looks ok.